sim/midas/src/main/scala/midas/passes/AutoCounterCoverTransform.scala

```diff
@@ -0,0 +1,411 @@
+import freechips.rocketchip.config.{Parameters, Field}
+import midas.widgets._
+import midas.targetutils._
+
+import java.io._
+import scala.io.Source
+import collection.mutable
+
+
+
+class FireSimPropertyLibrary() extends BasePropertyLibrary {
+  import chisel3._
+  import chisel3.experimental.DataMirror.internal.isSynthesizable
+  import chisel3.internal.sourceinfo.{SourceInfo}
+  import chisel3.core.{annotate,ChiselAnnotation}
+  def generateProperty(prop_param: BasePropertyParameters)(implicit sourceInfo: SourceInfo) {
+    //requireIsHardware(prop_param.cond, "condition covered for counter is not hardware!")
+    if (!(prop_param.cond.isLit) && chisel3.experimental.DataMirror.internal.isSynthesizable(prop_param.cond)) {
+      annotate(new ChiselAnnotation { def toFirrtl = AutoCounterCoverAnnotation(prop_param.cond.toNamed, prop_param.label, prop_param.message) })
+    }
+  }
+}
+//=========================================================================
+ 
+
+/**
+Take the annotated cover points and convert them to counters
+**/
+class AutoCounterCoverTransform(dir: File = new File("/tmp/"), printcounter: Boolean = false)(implicit p: Parameters) extends Transform {
+  def inputForm: CircuitForm = LowForm
+  def outputForm: CircuitForm = LowForm
+  override def name = "[FireSim] AutoCounter Cover Transform"
+  val newAnnos = mutable.ArrayBuffer.empty[Annotation]
+  val autoCounterLabels = mutable.ArrayBuffer.empty[String]
+  val autoCounterMods = mutable.ArrayBuffer.empty[Module]
+  val autoCounterLabelsSourceMap = mutable.Map.empty[String, String]
+  val autoCounterReadableLabels = mutable.ArrayBuffer.empty[String]
+  val autoCounterReadableLabelsMap = mutable.Map.empty[String, String]
+  val autoCounterPortsMap = mutable.Map.empty[Int, String]
+  val autoCounterLabelsMap = mutable.Map.empty[String, String]
+
+  private def MakeCounter(label: String, hastracerwidget: Boolean = false): CircuitState = {
+    import chisel3._
+    import chisel3.experimental.MultiIOModule
+    import chisel3.experimental.ChiselAnnotation
+    import chisel3.util.experimental.BoringUtils
+    def countermodule() = new MultiIOModule {
+      //override def desiredName = "AutoCounter"
+      val cycle = RegInit(0.U(64.W))
+      cycle := cycle + 1.U
+      val in0 = IO(Input(Bool()))
+
+      //connect the trigger from the tracer widget
+      val trigger = WireDefault(true.B)
+      hastracerwidget match {
+        case true => BoringUtils.addSink(trigger, s"trace_trigger")
+        case _ => trigger := true.B
+      }
+
+      //*****if we ever want to use the trigger to reset the counters******
+      //withReset(reset = ~trigger) {
+      //  val countfun = WideCounter(64, in0)
+      //}
+
+      val countfun = WideCounter(64, in0)
+      val count = Wire(UInt(64.W))
+      count := countfun
+      if (printcounter) {
+        when (in0 & trigger) {
+          printf(midas.targetutils.SynthesizePrintf(s"[AutoCounter] $label: %d cycle: %d\n",count, cycle))
```

Comment:
The printf widget already knows what cycle it is, so you can drop the free-running counter.

Reply:
The reason this is there is since we originally found some issues with the printf cycle count. Those were fixed, however, when I was testing this version, I found that the manual cycle counter counts from target reset, while the built-in printf cycle counter counts from simulation reset (hence - there is some consent offset between them). I was debating with myself whether this matter for any purpose, so I just left both of them. However, if you feel strongly, I can remove the target cycle counter.

Reply:
I removed the target cycle counter